### PR TITLE
assumeReadType for Union and Subset

### PR DESCRIPTION
**Pull request by rintcius**
* Elide `assumeReadType`-guards for `Union` and `Subset`
* Also started adding property tests for `assumeReadType`

#qz-3584 Done

### CODE REVIEW

File: mongodb/src/test/scala/quasar/physical/mongodb/planner/assumeReadTypeSpec.scala [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+import slamdata.Predef._
+
+import eu.timepit.refined.auto._
+import matryoshka.{Hole => _, _}
+import matryoshka.data._
+import org.scalacheck._
+import pathy.Path._
+import scalaz._
+
+class assumeReadTypeSpec extends Qspec with TTypes[Fix] with TreeMatchers {
```
Comment (jsantos17) [inline mongodb/src/test/scala/quasar/physical/mongodb/planner/assumeReadTypeSpec.scala]: Great to see property tests for `assumeReadType`!

**Comment by jsantos17** [comment]
Looks great. I couldn't break it easily :D

However, this query still has some typechecks on `LeftSide`:

```
(select city, state, loc[*] from extraSmallZips) union (select city, state, loc[*] from extraSmallZips)
```

<details>
<summary>
QScript with unnecessary typechecks
</summary>

```
Reduce
├─ Union
│  ├─ ShiftedRead(/jps/extraSmallZips, ExcludeId)
│  ├─ LeftBranch
│  │  ╰─ LeftShift
│  │     ├─ ○
│  │     ├─ Struct
│  │     │  ╰─ Guard
│  │     │     ├─ ProjectKey
│  │     │     │  ├─ ○
│  │     │     │  ╰─ Constant(Str(loc))
│  │     │     ├─ Type(FlexArr(0, None, Top))
│  │     │     ├─ ProjectKey
│  │     │     │  ├─ ○
│  │     │     │  ╰─ Constant(Str(loc))
│  │     │     ╰─ Undefined
│  │     ├─ IdStatus
│  │     │  ╰─ IdStatus(ExcludeId)
│  │     ├─ ShiftType
│  │     │  ╰─ ShiftType(Array)
│  │     ├─ OnUndefined
│  │     │  ╰─ OnUndefined(Emit)
│  │     ╰─ Repair
│  │        ╰─ ConcatMaps
│  │           ├─ ConcatMaps
│  │           │  ├─ MakeMap
│  │           │  │  ├─ Constant(Str(city))
│  │           │  │  ╰─ Guard
│  │           │  │     ├─ LeftSide
│  │           │  │     ├─ Type(Obj(Map[], Some(Top)))
│  │           │  │     ├─ ProjectKey
│  │           │  │     │  ├─ LeftSide
│  │           │  │     │  ╰─ Constant(Str(city))
│  │           │  │     ╰─ Undefined
│  │           │  ╰─ MakeMap
│  │           │     ├─ Constant(Str(state))
│  │           │     ╰─ Guard
│  │           │        ├─ LeftSide
│  │           │        ├─ Type(Obj(Map[], Some(Top)))
│  │           │        ├─ ProjectKey
│  │           │        │  ├─ LeftSide
│  │           │        │  ╰─ Constant(Str(state))
│  │           │        ╰─ Undefined
│  │           ╰─ MakeMap
│  │              ├─ Constant(Str(loc))
│  │              ╰─ RightSide
│  ╰─ RightBranch
│     ╰─ LeftShift
│        ├─ ○
│        ├─ Struct
│        │  ╰─ Guard
│        │     ├─ ProjectKey
│        │     │  ├─ ○
│        │     │  ╰─ Constant(Str(loc))
│        │     ├─ Type(FlexArr(0, None, Top))
│        │     ├─ ProjectKey
│        │     │  ├─ ○
│        │     │  ╰─ Constant(Str(loc))
│        │     ╰─ Undefined
│        ├─ IdStatus
│        │  ╰─ IdStatus(ExcludeId)
│        ├─ ShiftType
│        │  ╰─ ShiftType(Array)
│        ├─ OnUndefined
│        │  ╰─ OnUndefined(Emit)
│        ╰─ Repair
│           ╰─ ConcatMaps
│              ├─ ConcatMaps
│              │  ├─ MakeMap
│              │  │  ├─ Constant(Str(city))
│              │  │  ╰─ Guard
│              │  │     ├─ LeftSide
│              │  │     ├─ Type(Obj(Map[], Some(Top)))
│              │  │     ├─ ProjectKey
│              │  │     │  ├─ LeftSide
│              │  │     │  ╰─ Constant(Str(city))
│              │  │     ╰─ Undefined
│              │  ╰─ MakeMap
│              │     ├─ Constant(Str(state))
│              │     ╰─ Guard
│              │        ├─ LeftSide
│              │        ├─ Type(Obj(Map[], Some(Top)))
│              │        ├─ ProjectKey
│              │        │  ├─ LeftSide
│              │        │  ╰─ Constant(Str(state))
│              │        ╰─ Undefined
│              ╰─ MakeMap
│                 ├─ Constant(Str(loc))
│                 ╰─ RightSide
├─ Bucket
│  ╰─ List
│     ╰─ ○
├─ Reducers
│  ╰─ Arbitrary
│     ╰─ ○
╰─ Repair
   ╰─ ReduceIndex(\/-(0))
```

</details>

Adding a limit to both sides of the `union` eliminates the typechecks:

```
(select city, state, loc[*] from extraSmallZips limit 5) union (select city, state, loc[*] from extraSmallZips limit 5)
```

<details>
<summary>QScript with fixed typechecks</summary>

```
QScript Mongo (Prefer Projection):
Reduce
├─ Union
│  ├─ Map
│  │  ├─ Unreferenced
│  │  ╰─ Constant(Int(5))
│  ├─ LeftBranch
│  │  ╰─ Subset(Take)
│  │     ├─ Unreferenced
│  │     ├─ From
│  │     │  ╰─ LeftShift
│  │     │     ├─ ShiftedRead(/jps/extraSmallZips, ExcludeId)
│  │     │     ├─ Struct
│  │     │     │  ╰─ Guard
│  │     │     │     ├─ ProjectKey
│  │     │     │     │  ├─ ○
│  │     │     │     │  ╰─ Constant(Str(loc))
│  │     │     │     ├─ Type(FlexArr(0, None, Top))
│  │     │     │     ├─ ProjectKey
│  │     │     │     │  ├─ ○
│  │     │     │     │  ╰─ Constant(Str(loc))
│  │     │     │     ╰─ Undefined
│  │     │     ├─ IdStatus
│  │     │     │  ╰─ IdStatus(ExcludeId)
│  │     │     ├─ ShiftType
│  │     │     │  ╰─ ShiftType(Array)
│  │     │     ├─ OnUndefined
│  │     │     │  ╰─ OnUndefined(Emit)
│  │     │     ╰─ Repair
│  │     │        ╰─ ConcatMaps
│  │     │           ├─ ConcatMaps
│  │     │           │  ├─ MakeMap
│  │     │           │  │  ├─ Constant(Str(city))
│  │     │           │  │  ╰─ ProjectKey
│  │     │           │  │     ├─ LeftSide
│  │     │           │  │     ╰─ Constant(Str(city))
│  │     │           │  ╰─ MakeMap
│  │     │           │     ├─ Constant(Str(state))
│  │     │           │     ╰─ ProjectKey
│  │     │           │        ├─ LeftSide
│  │     │           │        ╰─ Constant(Str(state))
│  │     │           ╰─ MakeMap
│  │     │              ├─ Constant(Str(loc))
│  │     │              ╰─ RightSide
│  │     ╰─ Count
│  │        ╰─ Map
│  │           ├─ Unreferenced
│  │           ╰─ Constant(Int(5))
│  ╰─ RightBranch
│     ╰─ Subset(Take)
│        ├─ Unreferenced
│        ├─ From
│        │  ╰─ LeftShift
│        │     ├─ ShiftedRead(/jps/extraSmallZips, ExcludeId)
│        │     ├─ Struct
│        │     │  ╰─ Guard
│        │     │     ├─ ProjectKey
│        │     │     │  ├─ ○
│        │     │     │  ╰─ Constant(Str(loc))
│        │     │     ├─ Type(FlexArr(0, None, Top))
│        │     │     ├─ ProjectKey
│        │     │     │  ├─ ○
│        │     │     │  ╰─ Constant(Str(loc))
│        │     │     ╰─ Undefined
│        │     ├─ IdStatus
│        │     │  ╰─ IdStatus(ExcludeId)
│        │     ├─ ShiftType
│        │     │  ╰─ ShiftType(Array)
│        │     ├─ OnUndefined
│        │     │  ╰─ OnUndefined(Emit)
│        │     ╰─ Repair
│        │        ╰─ ConcatMaps
│        │           ├─ ConcatMaps
│        │           │  ├─ MakeMap
│        │           │  │  ├─ Constant(Str(city))
│        │           │  │  ╰─ ProjectKey
│        │           │  │     ├─ LeftSide
│        │           │  │     ╰─ Constant(Str(city))
│        │           │  ╰─ MakeMap
│        │           │     ├─ Constant(Str(state))
│        │           │     ╰─ ProjectKey
│        │           │        ├─ LeftSide
│        │           │        ╰─ Constant(Str(state))
│        │           ╰─ MakeMap
│        │              ├─ Constant(Str(loc))
│        │              ╰─ RightSide
│        ╰─ Count
│           ╰─ Map
│              ├─ Unreferenced
│              ╰─ Constant(Int(5))
├─ Bucket
│  ╰─ List
│     ╰─ ○
├─ Reducers
│  ╰─ Arbitrary
│     ╰─ ○
╰─ Repair
   ╰─ ReduceIndex(\/-(0))
```

</details>

It doesn't look like something easy to fix, but you probably know better.

File: mongodb/src/test/scala/quasar/physical/mongodb/planner/assumeReadTypeSpec.scala [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+
+  def nextFix(s: Free[fs.MongoQScript[Fix, ?], Hole]) = Gen.const(
+    fix.Subset(fix.Unreferenced, s, Take, free.Map(free.Unreferenced, func.Constant(json.int(1)))))
+
+  def nextFree(s: Free[fs.MongoQScript[Fix, ?], Hole]) = Gen.oneOf(
+    free.Sort(s, Nil, NonEmptyList((func.Hole, SortDir.Ascending))),
+    free.Subset(free.Unreferenced, s, Take, free.Map(free.Unreferenced, func.Constant(json.int(1)))),
+    free.Subset(s, free.Hole, Take, free.Map(free.Unreferenced, func.Constant(json.int(1)))))
+
+  def genRewriteSrcFree(maxDepth: Int = MAX_DEPTH): Gen[Free[fs.MongoQScript[Fix, ?], Hole]] =
```
[the remark this answers is not in the document]
Comment (rintcius) [reply mongodb/src/test/scala/quasar/physical/mongodb/planner/assumeReadTypeSpec.scala]: @jsantos17 See here for some problems and solutions regarding recursion (I used the maxDepth solution mentioned there): https://stackoverflow.com/questions/19829293/scalacheck-arbitrary-implicits-and-recursive-generators 

I am not seeing how `Cogen` can be used here. As far as I understand, `Cogen` can be used to generate functions properly. Let me know if I missed something.

**Comment by rintcius** [comment]
@jsantos17 Ah, thanks for that example query. Let's create a separate issue for that.

**Pull request merged**